docs/textsplitter.md

```diff
@@ -0,0 +1,284 @@
+* [Splitting algorithm](#splitting-algorithm)
+* [Recursive handling of oversized spans](#recursive-handling-of-oversized-spans)
+* [Cross-page boundary repair](#cross-page-boundary-repair)
+* [Chunk normalization](#chunk-normalization)
+* [Semantic overlap](#semantic-overlap)
+* [Examples](#examples)
+
+## High-level overview
+
+The `SentenceTextSplitter` is designed to:
+
+1. Produce semantically coherent chunks that align with sentence boundaries.
+2. Respect a maximum token count per chunk (hard limit of 500 tokens) plus a soft character length guideline (default 1,000 characters with a 20% overflow tolerance for merges / normalization). Size limit does not apply to figure blocks (chunks containing a `<figure>` may exceed the token limit; figures are never split).
+3. Keep structural figure placeholders (`<figure>...</figure>`) atomic: never split internally and always attach them to preceding accumulated text if any exists.
+4. Repair mid‑sentence page breaks when safe via merge or fragment shift heuristics while enforcing token + soft character budgets.
+5. Avoid empty outputs or unclosed figure tags.
+6. Perform a light normalization pass (trim only minimal leading/trailing whitespace that would cause small overflows; do not modify figure chunks).
+
+The splitter includes these components:
+
+* Segmentation based on sentence-ending punctuation ( `. ! ?` plus CJK equivalents ).
+* An internal accumulator that appends sentence‑like spans until the next addition would breach character or token limits, then flushes a chunk. Default hard token cap is 500 per chunk.
+* Recursive subdivision of oversized individual spans using a boundary preference order:
+    1. Sentence-ending punctuation near the midpoint (scan within the central third of the span).
+    2. If no sentence boundary is found, a word break (space / punctuation from a configured list) near the midpoint to avoid mid‑word cuts.
+    3. If neither boundary type is found, a symmetric 10% overlap midpoint split (duplicated region appears at the end of the first part and the start of the second) preserves continuity.
+* Figure handling is front‑loaded: figure blocks are extracted first and treated as atomic before any span splitting or recursion on plain text.
+* Cross‑page merge of text when all safety checks pass (prior chunk ends mid‑sentence, next chunk starts lowercase, not a heading, no early figure) and combined size fits both token and soft char budgets; otherwise a fragment shift may move the trailing unfinished clause forward.
+* A lightweight semantic overlap duplication pass (10% of max section length) that appends a trimmed prefix of the next chunk onto the end of the previous chunk (the next chunk itself is left unchanged). This is always attempted for adjacent non‑figure chunks on the same page and conditionally across a page boundary when the next chunk appears to be a direct lowercase continuation (and not a heading or figure). Figures are never overlapped/duplicated.
```

Comment:
rather than calling this "overlap duplication" - perhaps "adding overlap to each chunk"?
We may want to omit the "always attempted" sentence and just state "figures are never overlapped / duplicated"

Reply:
Sure, reworded